mahiru/components/network_administrator.py

```diff
@@ -86,3 +132,100 @@ def disconnect_inputs(self, job_id: int, inputs: Dict[str, Asset]) -> None:
             inputs: Input assets for this job.
         """
         raise NotImplementedError()
+
+    def _create_wg_endpoint(
+            self, pid: int, net: int, host: int) -> WireGuardEndpoint:
+        """Create a local WireGuard endpoint.
+
+        This will create an endpoint in the network namespace
+        with the given PID, for the given net and host.
+
+        Args:
+            pid: PID of the network namespace (container) to create the
+                endpoint inside of.
+            net: Network to create the endpoint for.
+            host: Host id, either _SERVER_HOST or _CLIENT_HOST.
+
+        Return:
+            The local endpoint for the remote side to connect to.
+        """
+        try:
+            our_port = self._available_ports.pop()
+        except KeyError:
+            raise RuntimeError('Insufficient resources')
+
+        try:
+            our_key = self._run_net_admin_helper(
+                    f'cwg_create {pid} {net} {host} {our_port}')
+        except Exception:
+            self._available_ports.add(our_port)
+            raise
+
+        our_ip = self._settings.external_ip
+        assert our_ip is not None
+        return WireGuardEndpoint(our_ip, our_port, our_key)
+
+    def _connect_wg_endpoint(
+            self, pid: int, net: int, host: int, endpoint: WireGuardEndpoint
+            ) -> None:
+        """Connect a local WireGuard endpoint to a remote one.
+
+        Args:
+            pid: PID of local namespace the endpoint is in.
+            net: The network the local endpoint (and hopefully the
+                    remote one too!) is for.
+            host: The local host id (either _SERVER_HOST or
+                    _CLIENT_HOST)
+            endpoint: The remote endpoint to connect to.
+        """
+        self._run_net_admin_helper(
+                f'cwg_connect {pid} {net} {host} {endpoint.endpoint()}'
+                f' {endpoint.key}')
+
+    def _remove_wg_endpoint(
+            self, pid: int, net: int, host: int) -> None:
+        """Remove a local WireGuard endpoint.
+
+        Args:
+            pid: PID of local namespace the endpoint is in.
+            net: The network the local endpoint is for.
+            host: The local host id (either _SERVER_HOST or
+                    _CLIENT_HOST)
+        """
+        self._run_net_admin_helper(f'cwg_destroy {pid} {net} {host}')
+
+    def _ensure_net_admin_helper(self) -> None:
+        """Ensures that the NAH Docker image is loaded into Docker."""
+        try:
+            self._dcli.images.get(_NAH_IMAGE)
+        except docker.errors.ImageNotFound:
+            image_file = (
+                    Path(__file__).parents[1] / 'data' /
+                    'net-admin-helper.tar.gz')
+            with image_file.open('rb') as f:
+                image = self._dcli.images.load(f.read())[0]
```

Comment:
I have seen similar code like this before. You could have a helper function: `ensure_image_loaded(image_path)` and reuse that throughout your codebase.

Reply:
Ah yes, for loading the pilot container. It's really only a few lines though, I'm not sure it's worth a `components/docker.py` with a separate function in it at this point.